source/mesh_deformation/diffusion.cc

```diff
@@ -0,0 +1,587 @@
+      // because the matrix is mostly empty, since we don't touch internal vertices.
+      LinearAlgebra::PreconditionJacobi preconditioner_mass;
+      LinearAlgebra::PreconditionJacobi::AdditionalData preconditioner_control(1,1e-16,1);
+      preconditioner_mass.initialize(matrix, preconditioner_control);
+
+      this->get_pcout() << "   Solving mesh surface diffusion" << std::endl;
+      SolverControl solver_control(5*system_rhs.size(), this->get_parameters().linear_stokes_solver_tolerance*system_rhs.l2_norm());
+      SolverCG<LinearAlgebra::Vector> cg(solver_control);
+      cg.solve (matrix, solution, system_rhs, preconditioner_mass);
+
+      // Distribute constraints on mass matrix
+      matrix_constraints.distribute (solution);
+
+      // The solution contains the new displacements, but we need to return a velocity.
+      // Therefore, we compute v=d_displacement/d_t.
+      // d_displacement are the new mesh node locations
+      // minus the old locations, which are initial_topography + displacements.
+      LinearAlgebra::Vector velocity(mesh_locally_owned, mesh_locally_relevant, this->get_mpi_communicator());
+      velocity = solution;
+      velocity -= initial_topography;
+      velocity -= displacements;
+
+      // The velocity
+      if (this->get_timestep() > 0.)
+        velocity /= this->get_timestep();
+      else
+        AssertThrow(false, ExcZero());
+
+      output = velocity;
+    }
+
+
+    template <int dim>
+    void Diffusion<dim>::check_diffusion_time_step (const DoFHandler<dim> &mesh_deformation_dof_handler,
+                                                    const std::set<types::boundary_id> &boundary_ids) const
+    {
+      // Initialize Gauss-Legendre quadrature for degree+1 quadrature points of the surface faces
+      const QGauss<dim-1> face_quadrature(mesh_deformation_dof_handler.get_fe().degree+1);
+
+      FEFaceValues<dim> fs_fe_face_values (this->get_mapping(), mesh_deformation_dof_handler.get_fe(), face_quadrature, update_default);
+
+      double min_local_conduction_timestep = std::numeric_limits<double>::max();
+
+      for (const auto &fscell : mesh_deformation_dof_handler.active_cell_iterators())
+        if (fscell->at_boundary() && fscell->is_locally_owned())
+          for (unsigned int face_no=0; face_no<GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (fscell->face(face_no)->at_boundary())
+              {
+                // Get the boundary indicator of current cell face
+                const types::boundary_id boundary_indicator
+                  = fscell->face(face_no)->boundary_id();
+
+                // Only consider the requested boundaries
+                if (boundary_ids.find(boundary_indicator) == boundary_ids.end())
+                  continue;
+
+                // Reinitialize update flags for current cell face
+                fs_fe_face_values.reinit (fscell, face_no);
```

Comment:
I do not see where you use `fs_fe_face_values`, can it be removed completetly?

Reply:
yes